### PR TITLE
Manage diagnostics from open files only analyzers better

### PR DESCRIPTION
**Pull request by heejaechang**
**Customer scenario**

customer opens a file and see diagnostics from open file only analyzers, but when he closes the file in VS, the diagnostics doesn't go away. that makes number of diagnostics increase as he opens and closes more files and no way to clear or update those until open the file again.

now, diagnostics from open file only analyzers will go away once user closes the file the diagnostics came from.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/15760

**Workarounds, if any**

no workaround

**Risk**

this fixes a bug where open file only diagnostics doesnt get cleared when file is closed. this behavior was design and current behavior was a bug. affected code by this fix is small so I believe risk is low.

**Performance impact**

code change is not something that can affect perf.

**Is this a regression from a previous update?**

no

**Root cause analysis:**

root cause of this is the fact that we have OpenFileOnly IDE analyzers. it would be nice if we can get rid of those by either making analyzer fast enough to run for all files or portable enough to allow it to run on OOP.

**How was the bug found?**

@davkean reported the issue.

### CODE REVIEW

**Comment by heejaechang** [comment]
@davkean @mavasani @srivatsn @panopticoncentral @CyrusNajmabadi can  you take a look?

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 1]
```diff
@@ -48,12 +48,31 @@ public static bool IsBuiltInAnalyzer(this DiagnosticAnalyzer analyzer)
...
-                return !builtInAnalyzer.OpenFileOnly(project.Solution.Workspace);
+                return builtInAnalyzer.OpenFileOnly(workspace);
+            }
+
+            return false;
+        }
+
+        public static bool ShouldRunForFullProject(this DiagnosticAnalyzerService service, DiagnosticAnalyzer analyzer, Project project)
+        {
+            if (analyzer.IsBuiltInAnalyzer())
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: I think the method name is confusing. will change it as IsAnalyzerWithNonHiddenDescriptor

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 2]
```diff
@@ -72,12 +72,13 @@ public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt,
             {
                 var stateSets = GetStateSetsForFullSolutionAnalysis(_stateManager.GetOrUpdateStateSets(project), project).ToList();
 
-                // PERF: get analyzers that are not suppressed.
+                // PERF: get analyzers that are not suppressed and marked as open file only
                 // this is perf optimization. we cache these result since we know the result. (no diagnostics)
                 // REVIEW: IsAnalyzerSuppressed call seems can be quite expensive in certain condition. is there any other way to do this?
                 var activeAnalyzers = stateSets
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs]: this is the one that decides whether the analyzer will actually run or not.

**Comment by heejaechang** [comment]
@MattGertz can I get approval for RC3?

**Comment by davkean** [comment]
You call out no behavior change above, but clearly this is a behavior change right? (which I believe is a good change, but a behavior change none-the-less).

**Comment by heejaechang** [comment]
@davkean I think you are confused with other PR? this does change behavior.

**Comment by heejaechang** [comment]
ah I see, updated the description. I put wrong description on risk.

**Comment by heejaechang** [comment]
@MattGertz can I get approval for RC3?

**Comment by MattGertz** [comment]
(Still waiting for one more reviewer approval.  Tag me when you get it.)

**Comment by heejaechang** [comment]
ping? can I get some code review?

**Comment by heejaechang** [comment]
@panopticoncentral can you take a look? @srivatsn

**Comment by panopticoncentral** [comment]
This looks OK to me, although I'm not as knowledgeable about this area.

**Comment by heejaechang** [comment]
@panopticoncentral @mavasani is probably only one except me who is familiar with this code.

so, can I assume you signed off?

**Comment by panopticoncentral** [comment]
:+1:

**Comment by heejaechang** [comment]
@MattGertz ping? I got 2 sign off

**Comment by MattGertz** [comment]
Approved.

**Pull request merged**